pkg/operator/controller/status.go

```diff
@@ -40,88 +47,167 @@ func (r *reconciler) syncOperatorStatus() error {
 		{
 			Resource: "namespaces",
 			Name:     "openshift-ingress-operator",
 		},
 		{
 			Resource: "namespaces",
-			Name:     ns.Name,
+			Name:     nsName,
 		},
 	}
-
-	// An available operator resets release version
-	for _, condition := range co.Status.Conditions {
-		if condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionTrue {
-			co.Status.Versions = []configv1.OperandVersion{
-				{
-					Name:    operatorVersionName,
-					Version: r.OperatorReleaseVersion,
-				},
-				{
-					Name:    "ingress-controller",
-					Version: r.RouterImage,
-				},
-			}
-		}
-	}
-
-	if !operatorStatusesEqual(*oldStatus, co.Status) {
-		err = r.client.Status().Update(context.TODO(), co)
-		if err != nil {
-			return fmt.Errorf("failed to update clusteroperator %s: %v", co.Name, err)
-		}
-	}
-
-	return nil
 }
 
 // getOperatorState gets and returns the resources necessary to compute the
 // operator's current state.
-func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.IngressController, error) {
-	ns := manifests.RouterNamespace()
-
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
+func (r *reconciler) getOperatorState(nsName string) ([]operatorv1.IngressController, *corev1.Namespace, error) {
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, ns); err != nil {
 		if errors.IsNotFound(err) {
 			return nil, nil, nil
 		}
-
-		return nil, nil, fmt.Errorf(
-			"error getting Namespace %s: %v", ns.Name, err)
+		return nil, nil, fmt.Errorf("error getting Namespace %s: %v", nsName, err)
 	}
 
 	ingressList := &operatorv1.IngressControllerList{}
 	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
 		return nil, nil, fmt.Errorf("failed to list IngressControllers: %v", err)
 	}
 
-	return ns, ingressList.Items, nil
+	return ingressList.Items, ns, nil
+}
+
+// computeOperatorStatusVersions computes the operator's current versions.
+func (r *reconciler) computeOperatorStatusVersions(oldVersions []configv1.OperandVersion, allIngressesAvailable bool) []configv1.OperandVersion {
+	oldReleaseVersion := UnknownVersionValue
+	oldIngressControllerImage := UnknownVersionValue
+	for _, opv := range oldVersions {
+		switch opv.Name {
+		case OperatorVersionName:
+			oldReleaseVersion = opv.Version
+		case IngressControllerVersionName:
+			oldIngressControllerImage = opv.Version
+		}
+	}
+
+	releaseVersion := r.OperatorReleaseVersion
+	ingressControllerImage := r.IngressControllerImage
+	// We need to report old version until the operator fully transitions to the new version.
+	// https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#version-reporting-during-an-upgrade
+	if !allIngressesAvailable {
+		releaseVersion = oldReleaseVersion
+		ingressControllerImage = oldIngressControllerImage
```

Comment:
This can be simplified to `return oldVersions` (and we don't need to compute `oldReleaseVersion`, `oldIngressControllerImage`, `releaseVersion`, or `ingressControllerImage`).